Move CommonCrypto and zlib dependencies mapping to custom module map

turbulem opened the pull request:
Hi!
I faced an issue similar to [this](https://github.com/daltoniam/Starscream/issues/445) and this [this](https://github.com/daltoniam/Starscream/issues/437).
In my project I use Carthage and commit my artefacts into my repo. And CI(and any other machine) is not able to compile it without running 'carthage update'.
Basically the problem that to link Starscream.framework with other app or library you should provide module.map similar to the one, located in /zlib folder (but without header, which is required only for build).
To make this work I created new custom module map and added a wrapper for SHA1 algorithm used by Starscream. Also updated cocoapods spec to add custom module map + private headers.

Please consider this PR as a possible way of resolving issues linked above and also I'd be happy to hear from you any suggestions on this.

Thank you!
Cheers!

kotsaarev commented:
Hello @turbulem.

Thank you for your solution! But I still have a problem.

I use GitLab CI, two build machines, cache server and Carthage. When I create cache on the first machine, and then use it on the second one, I get an error

```
Missing required modules: 'SSCZLib', 'SSCommonCrypto'
```

turbulem commented:
@kotsaarev Thank you for feedback. Let me check

turbulem commented:
I double checked. You're right. Looks like it actually doesn't solve an issue.
I have an alternative solution and will update this PR with it later this week.

JeanAzzopardi commented:
Just to add my 2c, I'm getting the same issue with Starscream on CircleCI - where I'm building carthage dependencies on a different lane and then using the cached dependencies for the test suite, and I get the same error: `missing required modules: 'SSCZLib', 'SSCommonCrypto'`

turbulem commented:
@JeanAzzopardi Did you check the most recent version?
The latest version in my repo should work fine:
https://github.com/turbulem/Starscream/commit/8ca8df0edc39ea1f2e25a827d0bdd998530755c5

JeanAzzopardi commented:
I haven't checked your version @turbulem, will try it out now and let you know how it goes.

JeanAzzopardi commented:
@turbulem - I can confirm your solution works on CircleCI

turbulem commented:
@JeanAzzopardi Thank you! Hope this PR will help you to fix the issue

JeanAzzopardi commented:
However, I just tested on device, and I am getting a crash:
```
 *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[NSConcreteData _fastCStringContents:]: unrecognized selector sent to instance 0x1c4237600'
```

turbulem commented:
@JeanAzzopardi Does this happen only on real device? Let me check later today on device. Thank you for your feedback!

JeanAzzopardi commented:
I will check and get back to you. Could be related to SSL connection too

On Wed, 27 Jun 2018 at 19:04, Sergey Lem <notifications@github.com> wrote:

> @JeanAzzopardi <https://github.com/JeanAzzopardi> Does this happen only
> on real device? Let me check later today on device. Thank you for your
> feedback!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/daltoniam/Starscream/pull/520#issuecomment-400757661>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAWzgiTy-pFXJ4MDFcYExx6t0kVnqJjks5uA7r_gaJpZM4Uob3u>
> .
>

turbulem commented:
@JeanAzzopardi It may be related to the ss_SHA1Base64Digest function that I added in my PR. This is small wrapper around CC, but all tests passed and I didn't check on device. I will have some time later this week or on a weekend.
But if you can check, that would be wonderful!

JeanAzzopardi commented:
@turbulem So I downloaded your branch and checked it out, the tests do pass, but trying the SimpleExample with the local websocket server on simulator gives me the same error.

turbulem commented:
Actually yes, I made a mistake. I return NSData instead of NSString and somehow compiler let me get away with it. Interesting...
I fixed it and updated PR.
Silly mistake, now it works well.
Thank you for your efforts to help!

JeanAzzopardi commented:
Yes, it's working now, that was the issue!

No problem, glad to help out, looking forward for this P.R. to be merged.

JeanAzzopardi commented:
@turbulem @daltoniam would it be possible to merge this? It's the last thing holding me back from upgrading to Starscream 3 <3

turbulem commented:
I've updated my branch with new changes

kotsaarev commented:
Hello @turbulem!
I think the problem with sha1Base64 still here, but occurs only in DEBUG mode. I checked out bd0732e and
```
*** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[Swift._NSContiguousString ss_SHA1Base64Digest]: unrecognized selector sent to instance 0x108b21050'
```
Crash is occurring in same lines
```
private extension String {
    func sha1Base64() -> String {
        return (self as NSString).ss_SHA1Base64Digest()
    }
}
```
I tested with iOS 11.4, Xcode 9.4.1, Swift 4.1. For more info see [crash file](https://github.com/daltoniam/Starscream/files/2240943/30-07-2018.crash.txt).

turbulem commented:
@kotsaarev  I will double check later this week. Thank you.

turbulem commented:
@kotsaarev So the final version should work fine. Let me know if I can help you anyhow.

dive commented:
@kotsaarev, any chance that it will be merged in the near future?

JeanAzzopardi commented:
@daltoniam Any chance that this could be merged? It's the only blocker preventing me from releasing an update as I use CircleCI for releases.

czater commented:
any chance to merge?

gregiOS commented:
Two things are missing here:
NSString+SHA1.m need to be added into Starscream (and test target)
<img width="1330" alt="screen shot 2018-09-06 at 12 08 23" src="https://user-images.githubusercontent.com/9901474/45151059-9d48b500-b1cd-11e8-83b6-51c0c9b5b185.png">
Other linkers flag need to contains `-all-load`
<img width="510" alt="screen shot 2018-09-06 at 12 09 23" src="https://user-images.githubusercontent.com/9901474/45151096-b6516600-b1cd-11e8-8239-9668dc630a16.png">

turbulem commented:
@gregiOS  Thank you for review. I made necessary changes. Now should be fine.

gregiOS commented:
Any chance for rebase @turbulem, and merge @daltoniam?

fassko commented:
@turbulem Can you please take a look on these merge conflicts. It will be easier to merge these changes. Thanks!

turbulem commented:
Hi, I updated my branch, however there are some issues with new Swift and tests need to be updated. I will do this later this week.

fassko commented:
OK let me know when I can check again.

fassko commented:
@turbulem please check:
* **SSLClientCertificate.swift** included twice in Sources folder
* Xcode project doesn't have files from Sources/Starscream to build

fassko commented:
File **SSLClientCertificate.swift** is in Sources and Sources/Starscream. I guess it should be just once in project. Thanks!

fassko commented on Package.swift:
```diff
@@ -21,12 +23,15 @@
+                .library(name: "Starscream", targets: ["Starscream"])
+        ],
+        dependencies: [
+                .package(url: "https://github.com/daltoniam/zlib-spm.git", from: "1.1.0"),
```
I'm being picky here, but for consistency let's use same format without **.git**

turbulem replied:
This is some merge issue, I will update my branch and discards this changes. 
I had no intention to change SPM config

turbulem replied:
Looks like it's some GH issue, because these files are equal:

https://github.com/daltoniam/Starscream/blob/master/Package.swift
https://github.com/turbulem/Starscream/blob/master/Package.swift

I will make another branch and refresh this PR in order to make it clear.

fassko replied:
OK, I can adjust it later myself.

turbulem commented:
@fassko Thank you for your instant reviews 👍

fassko commented:
@turbulem Let me know when you're done and I will merge this.

turbulem commented:
@fassko I opened new PR, please take a look.
